expression/builtin_time.go

```diff
@@ -1027,35 +1044,36 @@ func (c *yearFunctionClass) getFunction(args []Expression, ctx context.Context)
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinYearSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpTime)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen, bf.tp.Decimal = 4, 0
+	sig := &builtinYearSig{baseIntBuiltinFunc{bf}}
 	return sig.setSelf(sig), nil
 }
 
 type builtinYearSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinYearSig.
+// evalInt evals YEAR(date).
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_year
-func (b *builtinYearSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinYearSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	date, isNull, err := b.args[0].EvalTime(row, sc)
 	if err != nil {
-		return d, errors.Trace(err)
+		sc.AppendWarning(err)
 	}
-	d, err = convertToTime(b.ctx.GetSessionVars().StmtCtx, args[0], mysql.TypeDate)
-	if err != nil || d.IsNull() {
-		return d, errors.Trace(err)
+	if isNull || err != nil {
```

Comment:
in mysql:
```sql
MySQL > create table t(a bigint);
Query OK, 0 rows affected (0.02 sec)

MySQL > insert into t select year("aa");
ERROR 1292 (22007): Incorrect datetime value: 'aa'
MySQL > select year("aa");
+------------+
| year("aa") |
+------------+
|       NULL |
+------------+
1 row in set, 1 warning (0.00 sec)
```

seems mysql will throw this error during insert/update/delete, it will be solved in #4208